cheroot/ssl/pyopenssl.py

```diff
@@ -150,14 +150,26 @@ def readline(self, size=-1):
             size,
         )
 
-    def sendall(self, *args, **kwargs):
-        """Send whole message to the socket."""
+    def read(self, *args, **kwargs):
+        """Read from the wrapped socket, with retry."""
         return self._safe_call(
-            False,
-            super(SSLFileobjectMixin, self).sendall,
+            True,
+            super(SSLFileobjectMixin, self).read,
             *args, **kwargs
         )
 
+    def sendall(self, *args, **kwargs):
+        """Send whole message to the socket.
+
+        Not supported due to https://github.com/pyca/pyopenssl/issues/176.
+        Until that bug is fixed, sendall() may throw SSL.WantWriteError, but
+        there is no correct way to retry the call because we don't know how
+        many bytes were already transmitted. We could work around this by
+        reimplementing sendall() using send(), but we don't actually use
+        sendall() anywhere.
+        """
+        raise NotImplementedError('sendall() is unsupported by pyOpenSSL')
```

Comment:
How about taking https://github.com/pyca/pyopenssl/blob/124a013/src/OpenSSL/SSL.py#L1649-L1677 and adding retries and then resubmitting the fix to the upstream too?

Reply:
Another interesting code snippet: https://github.com/Lawouach/WebSocket-for-Python/issues/174#issuecomment-150142032.